django/forms/models.py

```diff
@@ -1176,6 +1177,11 @@ def choice(self, obj):
             self.field.label_from_instance(obj),
         )
 
+    def distinct_choices(self, queryset):
+        """Remove duplicates when using limit_choices_to."""
+        for obj in dict.fromkeys([_q for _q in queryset]):
```

Comment:
Why the obscure `dict.fromkeys` with a list comprehension that could be just `list(queryset)`?
IMHO the clearest code that wouldn't require a new method for clarification would be... 
```python
for obj in set(queryset):
   ...
```

If you feel the need to clarify why the `set` is used, maybe be more verbose in the unittest and/or commit message and squash the commits into one neat commit.

Reply:
`list(queryset)` is certainly better than what I was doing. I'll change that.

I also thought using `set` would be easier, but for some reason it fails when you run the entire test.
This works fine
```python
for obj in dict.fromkeys(list(queryset):
    ...
```
and even this works fine
```python
for obj in list(dict.fromkeys(list(queryset)):
    ...
```
but this fails
```python
for obj in set(dict.fromkeys(list(queryset)):
    ...
```
It seems like it doesn't get along with `set`. So I just went with what works.

Reply:
Do you have a traceback or the name(s) of test that fails with the set(). I ran the sqlite suite with the set() without problems.

Reply:
It's quite long. I had 6 failures.
```
======================================================================
FAIL: test_queryset_override (modeladmin.tests.ModelAdminTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.6/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/usr/lib/python3.6/unittest/case.py", line 605, in run
    testMethod()
  File "/home/alvin/git/django/tests/modeladmin/tests.py", line 480, in test_queryset_override
    '</select></div>' % (band2.id, self.band.id)
  File "/home/alvin/git/django/django/test/testcases.py", line 795, in assertHTMLEqual
    self.fail(self._formatMessage(msg, standardMsg))
  File "/usr/lib/python3.6/unittest/case.py", line 670, in fail
    raise self.failureException(msg)
AssertionError: <div class="related-widget-wrapper">
<select id="id_main_band" name="main_band"  [truncated]... != <div class="related-widget-wrapper">
<select id="id_main_band" name="main_band"  [truncated]...
  <div class="related-widget-wrapper">
  <select id="id_main_band" name="main_band" required>
  <option selected value>
  ---------
+ </option><option value="2">
+ The Beatles
  </option><option value="1">
  The Doors
- </option><option value="2">
- The Beatles
  </option>
  </select>
  </div>

======================================================================
FAIL: test_onetoonefield (model_forms.tests.ModelOneToOneFieldTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.6/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/usr/lib/python3.6/unittest/case.py", line 605, in run
    testMethod()
  File "/home/alvin/git/django/tests/model_forms/tests.py", line 1948, in test_onetoonefield
    self.w_woodward.pk, self.w_royko.pk,
  File "/home/alvin/git/django/django/test/testcases.py", line 795, in assertHTMLEqual
    self.fail(self._formatMessage(msg, standardMsg))
  File "/usr/lib/python3.6/unittest/case.py", line 670, in fail
    raise self.failureException(msg)
AssertionError: <p>
<label for="id_writer">
Writer:
</label><select id="id_writer" name="writer" [truncated]... != <p>
<label for="id_writer">
Writer:
</label><select id="id_writer" name="writer" [truncated]...
  <p>
  <label for="id_writer">
  Writer:
  </label><select id="id_writer" name="writer" required>
  <option selected value>
  ---------
+ </option><option value="2">
+ Bob Woodward
  </option><option value="1">
  Mike Royko
- </option><option value="2">
- Bob Woodward
  </option>
  </select>
  </p><p>
  <label for="id_age">
  Age:
  </label><input id="id_age" min="0" name="age" required type="number">
  </p>

======================================================================
FAIL: test_initial_values (model_forms.tests.ModelFormBasicTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.6/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/usr/lib/python3.6/unittest/case.py", line 605, in run
    testMethod()
  File "/home/alvin/git/django/tests/model_forms/tests.py", line 1264, in test_initial_values
    </select></li>''' % (self.w_woodward.pk, self.w_royko.pk, self.c1.pk, self.c2.pk, self.c3.pk))
  File "/home/alvin/git/django/django/test/testcases.py", line 795, in assertHTMLEqual
    self.fail(self._formatMessage(msg, standardMsg))
  File "/usr/lib/python3.6/unittest/case.py", line 670, in fail
    raise self.failureException(msg)
AssertionError: <li>
Headline:<input maxlength="50" name="headline" required type="text" value=" [truncated]... != <li>
Headline:<input maxlength="50" name="headline" required type="text" value=" [truncated]...
Diff is 1029 characters long. Set self.maxDiff to None to see it.

======================================================================
FAIL: test_multi_fields (model_forms.tests.ModelFormBasicTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.6/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/usr/lib/python3.6/unittest/case.py", line 605, in run
    testMethod()
  File "/home/alvin/git/django/tests/model_forms/tests.py", line 1428, in test_multi_fields
    </select></td></tr>''' % (self.w_woodward.pk, self.w_royko.pk, self.c1.pk, self.c2.pk, self.c3.pk))
  File "/home/alvin/git/django/django/test/testcases.py", line 795, in assertHTMLEqual
    self.fail(self._formatMessage(msg, standardMsg))
  File "/usr/lib/python3.6/unittest/case.py", line 670, in fail
    raise self.failureException(msg)
AssertionError: <tr>
<th>
Headline:
</th><td>
<input maxlength="50" name="headline" required typ [truncated]... != <tr>
<th>
Headline:
</th><td>
<input maxlength="50" name="headline" required typ [truncated]...
  <tr>
  <th>
  Headline:
  </th><td>
  <input maxlength="50" name="headline" required type="text">
  </td>
  </tr><tr>
  <th>
  Slug:
  </th><td>
  <input maxlength="50" name="slug" required type="text">
  </td>
  </tr><tr>
  <th>
  Pub date:
  </th><td>
  <input name="pub_date" required type="text">
  </td>
  </tr><tr>
  <th>
  Writer:
  </th><td>
  <select name="writer" required>
  <option selected value>
  ---------
+ </option><option value="2">
+ Bob Woodward
  </option><option value="1">
  Mike Royko
- </option><option value="2">
- Bob Woodward
  </option>
  </select>
  </td>
  </tr><tr>
  <th>
  Article:
  </th><td>
  <textarea cols="40" name="article" required rows="10">
  </td>
  </tr><tr>
  <th>
  Categories:
  </th><td>
  <select multiple name="categories">
  <option value="1">
  Entertainment
  </option><option value="2">
  It's a test
  </option><option value="3">
  Third test
  </option>
  </select>
  </td>
  </tr><tr>
  <th>
  Status:
  </th><td>
  <select name="status">
  <option selected value>
  ---------
  </option><option value="1">
  Draft
  </option><option value="2">
  Pending
  </option><option value="3">
  Live
  </option>
  </select>
  </td>
  </tr>

======================================================================
FAIL: test_runtime_choicefield_populated (model_forms.tests.ModelFormBasicTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.6/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/usr/lib/python3.6/unittest/case.py", line 605, in run
    testMethod()
  File "/home/alvin/git/django/tests/model_forms/tests.py", line 1602, in test_runtime_choicefield_populated
    </select></li>''' % (self.w_woodward.pk, self.w_royko.pk, self.c1.pk, self.c2.pk, self.c3.pk))
  File "/home/alvin/git/django/django/test/testcases.py", line 795, in assertHTMLEqual
    self.fail(self._formatMessage(msg, standardMsg))
  File "/usr/lib/python3.6/unittest/case.py", line 670, in fail
    raise self.failureException(msg)
AssertionError: <li>
Headline:<input maxlength="50" name="headline" required type="text">
</li>< [truncated]... != <li>
Headline:<input maxlength="50" name="headline" required type="text">
</li>< [truncated]...
  <li>
  Headline:<input maxlength="50" name="headline" required type="text">
  </li><li>
  Slug:<input maxlength="50" name="slug" required type="text">
  </li><li>
  Pub date:<input name="pub_date" required type="text">
  </li><li>
  Writer:<select name="writer" required>
  <option selected value>
  ---------
+ </option><option value="2">
+ Bob Woodward
  </option><option value="1">
  Mike Royko
- </option><option value="2">
- Bob Woodward
  </option>
  </select>
  </li><li>
  Article:<textarea cols="40" name="article" required rows="10">
  </li><li>
  Categories:<select multiple name="categories">
  <option value="1">
  Entertainment
  </option><option value="2">
  It's a test
  </option><option value="3">
  Third test
  </option>
  </select>
  </li><li>
  Status:<select name="status">
  <option selected value>
  ---------
  </option><option value="1">
  Draft
  </option><option value="2">
  Pending
  </option><option value="3">
  Live
  </option>
  </select>
  </li>

======================================================================
FAIL: test_foreignkeys_which_use_to_field (model_forms.tests.OtherModelFormTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.6/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/usr/lib/python3.6/unittest/case.py", line 605, in run
    testMethod()
  File "/home/alvin/git/django/tests/model_forms/tests.py", line 2511, in test_foreignkeys_which_use_to_field
    (22, 'Pear')))
  File "/usr/lib/python3.6/unittest/case.py", line 829, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/usr/lib/python3.6/unittest/case.py", line 1039, in assertTupleEqual
    self.assertSequenceEqual(tuple1, tuple2, msg, seq_type=tuple)
  File "/usr/lib/python3.6/unittest/case.py", line 1010, in assertSequenceEqual
    self.fail(msg)
  File "/usr/lib/python3.6/unittest/case.py", line 670, in fail
    raise self.failureException(msg)
AssertionError: Tuples differ: (('', '---------'), (<django.forms.models.ModelChoiceIteratorV[202 chars]re')) != (('', '---------'), (86, 'Apple'), (87, 'Core'), (22, 'Pear'))

First differing element 2:
(<django.forms.models.ModelChoiceIteratorV[34 chars]ear')
(87, 'Core')

+ (('', '---------'), (86, 'Apple'), (87, 'Core'), (22, 'Pear'))
- (('', '---------'),
-  (<django.forms.models.ModelChoiceIteratorValue object at 0x7f48eefb73c8>,
-   'Apple'),
-  (<django.forms.models.ModelChoiceIteratorValue object at 0x7f48eefb71d0>,
-   'Pear'),
-  (<django.forms.models.ModelChoiceIteratorValue object at 0x7f48ec938390>,
-   'Core'))

----------------------------------------------------------------------
Ran 14097 tests in 368.188s

FAILED (failures=6, skipped=1019, expected failures=4)
Destroying test database for alias 'default'...
Destroying test database for alias 'default'...
Destroying test database for alias 'default'...
Destroying test database for alias 'default'...
Destroying test database for alias 'default'...
Destroying test database for alias 'other'...
Destroying test database for alias 'other'...
Destroying test database for alias 'other'...
Destroying test database for alias 'other'...
Destroying test database for alias 'other'...
```

Reply:
Did you run the entire test, `./runtests.py`?

Reply:
I did.
I see what's going wrong, a set has no defined order. So the order of the select options is not guaranteed. But if I'm not mistaken, neither dicts have no defined order either. Maybe there's something in itertools you can use. Otherwise collections.OrderdDict.fromkeys will guarantee the order is kept.

Reply:
Django assumes dicts are ordered so that should work. 

Django also has an ordered set which may help you.

https://www.github.com/django/django/tree/master/django%2Futils%2Fdatastructures.py

Reply:
I also thought it was the order thing, that's why I used `dict.fromkeys`. I tried using set with just one object, where order doesn't matter, but it still fails.
I've also found out that `dict.fromkeys` is only ordered in Python 3.6 and above, so I've change it to `collections.OrderdDict.fromkeys`, which also doesn't need converting the queryset to list.

Reply:
+1 for the ordered set... It would fit right in the original iteration without the extra method. If you do insist on keeping the extra method, consider prefixing it with a '_' to denote it is only used in this class. When I approach large codebases I like it when I can just see the intent of code and not have to switch from reading to a Silver Searcher search to guess from where it might be called.